src/device/log.cpp

```diff
@@ -0,0 +1,136 @@
+#include "misc_log_ex.h"
+#include "log.hpp"
+
+namespace hw {
+    namespace ledger {
+
+    void buffer_to_str(char *to,  const char *buff, int len) {
+      for (int i=0; i<len; i++) {
+        sprintf(to+2*i, "%.02x", (unsigned char)buff[i]);
+      }
+    }
+    void bufferLE_to_str(char *to,  const char *buff, int len) {
+      for (int i=0; i<len; i++) {
+        sprintf(to+2*i, "%.02x", (unsigned char)buff[31-i]);
+      }
+    }
+
+    void log_hexbuffer(std::string msg,  const char* buff, int len) {
+      char logstr[1024];
+      buffer_to_str(logstr,  buff, len);
+      MCDEBUG("ledger", msg<< ": " << logstr);
+    }
+
+    void log_hexbufferLE(std::string msg,  const char* buff, int len) {
+      char logstr[1024];
+      bufferLE_to_str(logstr,  buff, len);
+      MCDEBUG("ledger", msg<< ": " << logstr);
+    }
+
+    void log_message(std::string msg,  std::string info ) {
+      MCDEBUG("ledger", msg << ": " << info);
+    }
+
+    #ifdef DEBUGLEDGER
+    extern crypto::secret_key viewkey;
+    extern crypto::secret_key spendkey;
+
+
+    void decrypt(char* buf, int len) {
+      #ifdef IODUMMYCRYPT
+      int i;
+      if (len == 32) {
+        //view key?
+        for (i = 0; i<32; i++) {
+          if (buf[i] != 0) break;
+        }
+        if (i == 32) {
+          memmove(buf, hw::ledger::viewkey.data, 32);
+          return;
+        }
+        //spend key?
+        for (i = 0; i<32; i++) {
+          if (buf[i] != (char)0xff) break;
+        }
+        if (i == 32) {
+          memmove(buf, hw::ledger::spendkey.data, 32);
+          return;
+        }
+      }
+      //std decrypt: XOR.55h
+      for (i = 0; i<len;i++) {
+          buf[i] ^= 0x55;
+        }
+      #endif
+    }
```

Comment:
What is this for ?

Reply:
Again this is to debug whiout be totally blind. 

In prod the device encrypt a with AES all sensitive data that leave the device. 

In DEBUGLEDGER it exist two other mode:

In Dummy mode, the encryption is only a xor with 55h, allowing the device to retrieve the value and control the computation according to default device

In nocrypt all is clear.

Again this is only debug mode. Prod compiles without DEBUGLEDER activated :)